envoy/router/router.h

```diff
@@ -581,6 +566,16 @@ class VirtualHost {
    *         rather than no limit applies.
    */
   virtual uint32_t retryShadowBufferLimit() const PURE;
+
+private:
+  friend class RouteEntryImplBase;
```

Comment:
@mattklein123 this may not what you want, it is required by https://github.com/envoyproxy/envoy/pull/17449/files/53a962d6e6d34bb91ffb3937dd90da95939a4e14..202bf5655ae666a93bf2ce55adcf0cb55918d89d#diff-caa28d598c49a80fee49bb96c71783c953c6840fb72821cebfc2faff22d49563R600

Reply:
Once you are inside the impl, can you just use impl methods? I'm not sure why this has to be defined at the interface level?

Reply:
oh, you are right. It needn't to be in interface. I thought there are a lot of subclasses of Route, then we need the virtual functional here. But actually, it is not.

Reply:
There are two things.

First, the `RouteEntryImplBase::mostSpecificPerFilterConfig` is using VirtualHost's perFilterConfig interface which I can't remove from the VirtualHost interface.
https://github.com/envoyproxy/envoy/pull/17449/files#diff-caa28d598c49a80fee49bb96c71783c953c6840fb72821cebfc2faff22d49563R598-R601

Second, `RouteEntryImplBase::traversePerFilterConfig` it needs the RouteEntry's perFilterConfig.  
https://github.com/envoyproxy/envoy/pull/17449/files#diff-2663fc9624b6710b28f768baef19fe0d321f541c74be11a492735ae90bd681cdR1026-R1031

From the name of variable `maybe_weighted_cluster_config`, it should be for `WeightedClusterEntry`. But I think I can remove that, since `WeightedClusterEntry` has the own implement of `traversePerFilterConfig`

https://github.com/envoyproxy/envoy/pull/17449/files#diff-2663fc9624b6710b28f768baef19fe0d321f541c74be11a492735ae90bd681cdR1080-R1090

Reply:
I think all usage is inside config_impl though, right? Can you just replace calls to virtualHost() with a direct access to the impl class? I'm pretty sure this can all be pretty easily refactored to be internal to config_impl only and not the interface?

Reply:
oh, right, I can use the member directly since it is virtualHostImpl, let me update

Reply:
ah, you are right, I got it, let me update.